[AIRFLOW-6414] configuration docs (#6983)
mistercrunch/Airflow: docs/howto/configurations-ref.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Configuration Reference
=======================

.. _config-ref/core:

``[core]``
^^^^^^^^^^

> dags_folder

Reviewer: Can you also add labels before keys?

Author: So add a label for every key in the document? I could probably make a vim shortcut for that but just want to make sure I'm understanding the ask first

Reviewer: Yes. I mean creating a label for each key so that we can include a link to each key in the documentation. If you want I can do it very quickly using a regular expression. ;-D

Reviewer: Here is  doc with sections: https://pastebin.com/aL7qbni1 (take it or throw it away ;-))

> dags_folder

Reviewer: ```suggestion
``dags_folder``
```
This will make reading easier because the text will be in a fixed width font.

store_serialized_dags
*********************

Whether to serialises DAGs and persist them in DB.  If set to True, Webserver reads from DB instead of parsing DAG files More details: https://airflow.apache.org/docs/stable/dag-serialization.html

min_serialized_dag_update_interval
**********************************

Updating serialized DAG can not be faster than a minimum interval to reduce database write rate.

check_slas
**********

On each dagrun check against defined SLAs

.. _config-ref/logging:

[logging]
^^^^^^^^^

base_log_folder
***************
The folder where airflow should store its log files This path must be absolute

remote_logging
**************
Airflow can store logs remotely in AWS S3, Google Cloud Storage or Elastic Search. Users must supply an Airflow connection id that provides access to the storage location. If remote_logging is set to true, see UPDATING.md for additional configuration requirements.

remote_log_conn_id
******************

remote_base_log_folder
**********************

encrypt_s3_logs
***************

logging_level
*************

fab_logging_level
*****************

> Logging class

Reviewer: It seems to me that there is no such configuration option. If this is to be an option grouping section then the options should be nested. In my opinion, we should not add new sections in the reference documentation. The reference documentation should be as simple as possible and therefore simpler to maintain.

*************

Specify the class that will specify the logging configuration
This class has to be on the python classpath

logging_config_class
********************

Log format
**********

Colour the logs when the controlling terminal is a TTY.

colored_console_log
*******************

colored_log_format
******************

colored_formatter_class
***********************

log_format
**********

simple_log_format
*****************

task_log_prefix_template
************************

Specify prefix pattern like mentioned below with stream handler TaskHandlerWithCustomFormatter

log_filename_template
*********************
Log filename format

log_processor_filename_template
*******************************

dag_processor_manager_log_location
**********************************

Name of handler to read task instance logs. Default to use task handler.

task_log_reader
***************

> cli

Reviewer: ```suggestion
``[cli]``
```

> ***

Reviewer: ```suggestion
^^^
```

If you set web_server_url_prefix, do NOT forget to append it here, ex: endpoint_url. So api will look like: http://localhost:8080/myroot/api/experimental/...

.. _config-ref/debug:

[debug]
^^^^^^^

fail_fast
*********
Used only with DebugExecutor. If set to True DAG will fail with first failed task. Helpful for debugging purposes.

.. _config-ref/api:

[api]
^^^^^
auth_backend
************
How to authenticate users of the API

lineage

backend
*******

what lineage backend to use

.. _config-ref/atlas:

[atlas]
^^^^^^^
sasl_enabled
************
host
****
port
****
username
********
password
********

.. _config-ref/operators:

[operators]
^^^^^^^^^^^

default_owner
*************
The default owner assigned to each new operator, unless provided explicitly or passed via ``default_args``

default_cpus
************
default_ram
***********
default_disk
************
default_gpus
************

allow_illegal_arguments
***********************
Is allowed to pass additional/unused arguments (args, kwargs) to the BaseOperator operator. If set to False, an exception will be thrown, otherwise only the console message will be displayed.

.. _config-ref/hive:

> [hive]

Reviewer: ```suggestion
``[hive]``
```

^^^^^^

default_hive_mapred_queue
*************************
Default mapreduce queue for HiveOperator tasks

> mapred_job_name_template

Reviewer: ```suggestion

mapred_job_name_template
```

************************
Template for mapred_job_name in HiveOperator, supports the following named parameters: hostname, dag_id, task_id, execution_date

.. _config-ref/webserver:

> [webserver]

Reviewer: ```suggestion
``[webserver]``
```

^^^^^^^^^^^

base_url
********
The base url of your website as airflow cannot guess what domain or cname you are using. This is used in automated emails that airflow sends to point links to the right web server

> web_server_host

Reviewer: ```suggestion

``web_server_host``
```

***************
The ip specified when starting the web server

web_server_port
***************

The port on which to run the web server

web_server_ssl_cert
*******************
Paths to the SSL certificate and key for the web server. When both are provided SSL will be enabled. This does not change the web server port.

web_server_ssl_key
******************

web_server_master_timeout
*************************
Number of seconds the webserver waits before killing gunicorn master that doesn't respond

web_server_worker_timeout
*************************

Number of seconds the gunicorn webserver waits before timing out on a worker

> worker_refresh_batch_size

Reviewer: ```suggestion

``worker_refresh_batch_size``
```

*************************
Number of workers to refresh at a time. When set to 0, worker refresh is disabled. When nonzero, airflow periodically refreshes webserver workers by bringing up new ones and killing old ones.

worker_refresh_interval
***********************

Number of seconds to wait before refreshing a batch of workers.

> secret_key

Reviewer: ```suggestion

secret_key
```

**********

Secret key used to run your flask app. It should be as random as possible
workers
*******

> 

Reviewer: ```suggestion
```

Number of workers to run the Gunicorn web server

worker_class
************
The worker class gunicorn should use. Choices include sync (default), eventlet, gevent

> access_logfile

Reviewer: ```suggestion
``access_logfile``
```

**************

> 

Reviewer: ```suggestion
```

Log files for the gunicorn webserver. '-' means log to stderr.

error_logfile
*************

expose_config
*************
Expose the configuration file in the web server

expose_hostname
***************

> 

Reviewer: ```suggestion
```

Expose hostname in the web server

expose_stacktrace
*****************

> 

Reviewer: ```suggestion
```

Expose stacktrace in the web server

dag_default_view
****************

> 

Reviewer: ```suggestion
```

Default DAG view.  Valid values are: tree, graph, duration, gantt, landing_times

dag_orientation
***************

Default DAG orientation. Valid values are: LR (Left->Right), TB (Top->Bottom), RL (Right->Left), BT (Bottom->Top)

demo_mode
*********
Puts the webserver in demonstration mode; blurs the names of Operators for privacy.

log_fetch_timeout_sec
*********************
The amount of time (in secs) webserver will wait for initial handshake while fetching logs from other worker machine

hide_paused_dags_by_default
***************************
By default, the webserver shows paused DAGs. Flip this to hide paused DAGs by default

page_size
*********
Consistent page size across all listing views in the UI

navbar_color
************

> 

Reviewer: ```suggestion
```

Define the color of navigation bar

default_dag_run_display_number
******************************
Default dagrun to show in UI

enable_proxy_fix
****************
Enable werkzeug ``ProxyFix`` middleware

cookie_secure
*************
Set secure flag on session cookie

cookie_samesite
***************
Set samesite policy on session cookie

default_wrap
************
Default setting for wrap toggle on DAG code and TI log views.

analytics_tool
**************
Send anonymous user activity to your analytics tool

analytics_id
*************

update_fab_perms
****************
Update FAB permissions and sync security manager roles on webserver startup

force_log_out_after
*******************
Minutes of non-activity before logged out from UI 0 means never get forcibly logged out

.. _config-ref/email:

[email]
^^^^^^^

email_backend
*************

.. _config-ref/smtp:

[smtp]
^^^^^^

smtp_starttls
*************
If you want airflow to send emails on retries, failure, and you want to use the airflow.utils.email.send_email_smtp function, you have to configure an

smtp_ssl
********

> smtp_user

Reviewer: ```suggestion

smtp_user
```

*********
Uncomment and set the user/pass settings if you want to use SMTP AUTH

> smtp_password

Reviewer: ```suggestion
smtp_password

```

*************

> smtp_port

Reviewer: ```suggestion

smtp_port
```

*********

> smtp_mail_from

Reviewer: ```suggestion

smtp_mail_from
```

**************

.. _config-ref/sentry:

[sentry]
^^^^^^^^

> sentry_dsn

Reviewer: ```suggestion

sentry_dsn
```

flower_host
***********
Celery Flower is a sweet UI for Celery. Airflow has a shortcut to start it ``airflow flower``. This defines the IP that Celery Flower runs on

flower_url_prefix
*****************
The root URL for Flower Ex: flower_url_prefix

flower_port
***********
This defines the port that Celery Flower runs on

Securing Flower with Basic Authentication
Accepts user:password pairs separated by a comma
Example: flower_basic_auth
flower_basic_auth

default_queue
*************
Default queue that tasks get assigned to and that worker listen on.

sync_parallelism
****************
How many processes CeleryExecutor uses to sync task state. 0 means to use max(1, number of cores - 1) processes.

celery_config_options
*********************
Import path for celery configuration options

.. _config-ref/in_case_of_ssl :

> [In case of using SSL]

Reviewer: ```suggestion
```

^^^^^^^^^^^^^^^^^^^^^^
ssl_active
**********

ssl_key
*******

ssl_cacert
**********

> ssl_cert

Reviewer: ```suggestion

ssl_cert
```

********

.. _config-ref/celery_pool_imp :

> [Celery Pool implementation.]

Reviewer: ```suggestion
pool
```

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
****
```

See:
https://docs.celeryproject.org/en/latest/userguide/workers.html#concurrency
https://docs.celeryproject.org/en/latest/userguide/concurrency/eventlet.html
pool

.. _config-ref/celery_broker_transport_options:

[celery_broker_transport_options]
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
This section is for specifying options which can be passed to the
underlying celery broker transport.  See:
http://docs.celeryproject.org/en/latest/userguide/configuration.html#std:setting-broker_transport_options

visibility_timeout
******************

The visibility timeout defines the number of seconds to wait for the worker to acknowledge the task before the message is redelivered to another worker. Make sure to increase the visibility timeout to match the time of the longest ETA you're planning to use.

visibility_timeout is only supported for Redis and SQS celery brokers.
See:
http://docs.celeryproject.org/en/master/userguide/configuration.html#std:setting-broker_transport_options

.. _config-ref/dask:

[dask]
^^^^^^

This section only applies if you are using the DaskExecutor in
[core] section above

cluster_address
***************
The IP address and port of the Dask cluster's scheduler.

> tls_ca

Reviewer: ```suggestion

tls_ca
```

******
TLS/ SSL settings to access a secured Dask scheduler.

> tls_key

Reviewer: ```suggestion

tls_key
```

*******

> tls_cert

Reviewer: ```suggestion

tls_cert
```

num_runs
********
The number of times to try to schedule each DAG file
-1 indicates unlimited number

processor_poll_interval
***********************
The number of seconds to wait between consecutive DAG file processing

min_file_process_interval
*************************
after how much time (seconds) a new DAGs should be picked up from the filesystem

dag_dir_list_interval
*********************
How often (in seconds) to scan the DAGs directory for new files. Default to 5 minutes.

print_stats_interval
********************
How often should stats be printed to the logs

scheduler_health_check_threshold
********************************
If the last scheduler heartbeat happened more than scheduler_health_check_threshold ago (in seconds), scheduler is considered unhealthy. This is used by the health check in the "/health" endpoint

child_process_log_directory
***************************

scheduler_zombie_task_threshold
*******************************
Local task jobs periodically heartbeat to the DB. If the job has not heartbeat in this many seconds, the scheduler will mark the associated task instance as failed and will re-schedule the task.

catchup_by_default
******************

> max_tis_per_query

Reviewer: ```suggestion

max_tis_per_query
```

*****************
Turn off scheduler catchup by setting this to False. Default behavior is unchanged and Command Line Backfills still work, but the scheduler will not do scheduler catchup if this is False, however it can be set on a per DAG basis in the DAG definition (catchup)

This changes the batch size of queries in the scheduling main loop. If this is too high, SQL query performance may be impacted by one or more of the following:
- reversion to full table scan
- complexity of query predicate
- excessive locking

Additionally, you may hit the maximum allowable query length for your db.

Set this to 0 for no limit (not advised)

.. _config-ref/statsd:

> [statsd] (https://github.com/etsy/statsd) integration settings

Reviewer: ```suggestion
``[statsd]``
```
I would prefer to keep headers simple so that it is easier to navigate the entire document.  We can add the same link in the section description. I will try to add more crosslinks in the next PR when this PR is accepted.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> statsd_on

Reviewer: ```suggestion

statsd_on
```

*********

> statsd_host

Reviewer: ```suggestion

statsd_host
```

***********

> statsd_port

Reviewer: ```suggestion

statsd_port
```

***********

> statsd_prefix

Reviewer: ```suggestion

statsd_prefix
```

git_branch
**********

git_subpath
***********

git_sync_rev
************
The specific rev or hash the git_sync init container will checkout This becomes GIT_SYNC_REV environment variable in the git_sync init container for worker pods

git_user
********
Use git_user and git_password for user authentication or git_ssh_key_secret_name and git_ssh_key_secret_key for SSH authentication

git_password
************

git_sync_root
*************

git_sync_dest
*************

git_dags_folder_mount_point
***************************
Mount point of the volume if git-sync is being used. i.e. {AIRFLOW_HOME}/dags

To get Git-sync SSH authentication set up follow this format

.. code-block:: yaml

  airflow-secrets.yaml:
  ---
  apiVersion: v1
  kind: Secret
  metadata:
    name: airflow-secrets
    data:
      # key needs to be gitSshKey
      gitSshKey: <base64_encoded_data>
  ---
  airflow-configmap.yaml:
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: airflow-configmap
  data:
    known_hosts: |
      github.com ssh-rsa <...>
  airflow.cfg: |
  ...

git_ssh_key_secret_name
***********************

> git_ssh_known_hosts_configmap_name

Reviewer: ```suggestion

git_ssh_known_hosts_configmap_name
```

**********************************

> git_ssh_key_secret_name

Reviewer: ```suggestion

git_ssh_key_secret_name
```

***********************

> git_ssh_known_hosts_configmap_name

Reviewer: ```suggestion

git_ssh_known_hosts_configmap_name
```

**********************************

#TODO FIX CODE SAMPLE

To give the git_sync init container credentials via a secret, create a secret with two fields: GIT_SYNC_USERNAME and GIT_SYNC_PASSWORD (example below) and add `git_sync_credentials_secret

Secret Example:

.. code-block:: yaml

  apiVersion: v1
  kind: Secret
  metadata:
    name: git-credentials
  data:
    GIT_SYNC_USERNAME: <base64_encoded_git_username>
    GIT_SYNC_PASSWORD: <base64_encoded_git_password>
    git_sync_credentials_secret

For cloning DAGs from git repositories into volumes: https://github.com/kubernetes/git-sync

git_sync_container_repository
*****************************

> git_sync_container_tag

Reviewer: ```suggestion

git_sync_container_tag
```

**********************

> git_sync_init_container_name

Reviewer: ```suggestion

git_sync_init_container_name
```

****************************

> git_sync_run_as_user

Reviewer: ```suggestion

git_sync_run_as_user
```

image_pull_secrets
******************
Any image pull secrets to be given to worker pods, If more than one secret is required, provide a comma separated list: secret_a,secret_b

in_cluster
**********
Use the service account kubernetes gives to pods to connect to kubernetes cluster. It's intended for clients that expect to be running inside a pod running on kubernetes. It will raise an exception if called from a process not running in a kubernetes environment.

cluster_context
***************
When running with in_cluster options to Kubernetes client. Leave blank these to use default behaviour like ````kubectl```` has.

config_file
***********

affinity
********
Affinity configuration as a single line formatted JSON object. See the affinity model for top-level key names (e.g. ``nodeAffinity``, etc.): https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#affinity-v1-core

tolerations
***********
A list of toleration objects as a single line formatted JSON array See: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#toleration-v1-core

kube_client_request_args
************************

> **kwargs parameters to pass while calling a kubernetes client core_v1_api methods from Kubernetes Executor provided as a single line formatted JSON dictionary string. List of supported params in **kwargs are similar for all core_v1_apis, hence a single config variable for all apis See: https://raw.githubusercontent.com/kubernetes-client/python/master/kubernetes/client/apis/core_v1_api.py

Reviewer: ```suggestion
keywords parameters to pass while calling a kubernetes client core_v1_api methods from Kubernetes Executor provided as a single line formatted JSON dictionary string. List of supported keyworddsparams are similar for all core_v1_apis, hence a single config variable for all apis See: https://raw.githubusercontent.com/kubernetes-client/python/master/kubernetes/client/apis/core_v1_api.py
```
``**`` has special meaning in RST

> 

Reviewer: Can you add license header?

> 

Reviewer: Can you add page title e.g. "Configuration Reference"?  The title is very important because it is used to generate links in menu.

> Core

Reviewer: ```suggestion

.. _config-ref/core:

``[core]``
``````````
Can you add labels? This will allow us to refer to fragments more easily. We should use precise section names case-sensitive.
